### PR TITLE
Changing IRetainable.Release return type to bool & removing IsRetainable

### PR DESCRIPTION
**Pull request by ahsonkhan**
- Also removing the IsRetainable property

cc @KrzysztofCwalina, @davidfowl, @shiftylogic, @stephentoub

### CODE REVIEW

File: src/System.Buffers.Experimental/ReferenceCountedBuffer.cs [hunk 1]
```diff
@@ -17,15 +17,16 @@ public override void Retain()
...
         {
             if (!IsRetained) BuffersExperimentalThrowHelper.ThrowInvalidOperationException();
             if (Interlocked.Decrement(ref _referenceCount) <= 0) {
                 OnNoReferences();
             }
+            return IsRetained;
         }
 
-        public override bool IsRetained => _referenceCount > 0;
+        protected override bool IsRetained => _referenceCount > 0;
```
Comment (davidfowl) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: This change isn't right. The idea was to return the result of the atomic operation.

```C#
public override bool Release()
{
  var result = Interlocked.Decrement(ref _referenceCount);
  if (result < 0) BuffersExperimentalThrowHelper.ThrowInvalidOperationException();
  if (result == 0) {
      OnNoReferences();
      return true;
  }
  return false;
}
```

Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: @davidfowl, what do you mean by that? Are you proposing that we return the refcount instead of bool? If yes, we cannot do it. There are some lifetime management strategies/algorithms that don't relly know the exact refcount; they just know if the refcount is positive or zero.

Comment (davidfowl) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: I'm proposing that the ones that do ref count need to be written like what I have above so that it's correct. Ones that don't ref count I have no idea what it would look like (if we have any implementations like that besides the managed array).

Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: OK, I think I know what you mean. I was confused cause your comment is below the IsRetained property. Yes, we need to make the implementation of Release better.

Comment (shiftylogic) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: I see now you removed it from the IRetainable interface. What value is the property here?

File: src/System.Buffers.Primitives/System/Internal/OwnedArray.cs [hunk 2]
```diff
@@ -68,20 +68,21 @@ public override void Retain()
...
 
-        public override void Release()
+        public override bool Release()
         {
             if (!IsRetained) BufferPrimitivesThrowHelper.ThrowInvalidOperationException();
             if (Interlocked.Decrement(ref _referenceCount) == 0)
             {
                 OnNoReferences();
             }
+            return IsRetained;
```
Comment (shiftylogic) [inline src/System.Buffers.Primitives/System/Internal/OwnedArray.cs]: This is racy and has the potential for conflict. The interlocked operation above could bring the ref count to a value above 0 (i.e. this code does not do the disposed) but by the time this line of code runs, someone else did bring it down to 0. Then, both callers get a result of true. If the one that didn't end up releasing it makes some decision based on the belief that they did (via the return value), that could cause an issue.

Comment (ahsonkhan) [inline src/System.Buffers.Primitives/System/Internal/OwnedArray.cs]: Yep, agreed. This needs to be addressed.

File: src/System.Buffers.Experimental/ReferenceCountedBuffer.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: The property is on OwnedBuffer. It needs to know its own refcount, but it does not call Release on itself.

Comment (davidfowl) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: Why is the property there at all. It shouldn't be protected either.

Comment (ahsonkhan) [inline src/System.Buffers.Experimental/ReferenceCountedBuffer.cs]: The property is used in the OwnedBuffer.Dispose method. We want to throw an exception on Dispose when ref count is greater than 0.

**Comment by ahsonkhan** [comment]
@davidfowl, @shiftylogic - please review

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs [hunk 3]
```diff
@@ -106,15 +106,12 @@ static void Buffer(OwnedBuffer<byte> buffer)
         static void Pin(OwnedBuffer<byte> buffer)
         {
             var memory = buffer.Buffer;
-            Assert.False(buffer.IsRetained);
             var handle = memory.Retain(pin: true);
             unsafe
             {
                 Assert.NotEqual(0L, new IntPtr(handle.PinnedPointer).ToInt64());
             }
-            Assert.True(buffer.IsRetained);
```
Comment (KrzysztofCwalina) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: Why is this line removed?

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs [hunk 4]
```diff
@@ -106,15 +106,12 @@ static void Buffer(OwnedBuffer<byte> buffer)
         static void Pin(OwnedBuffer<byte> buffer)
         {
             var memory = buffer.Buffer;
-            Assert.False(buffer.IsRetained);
```
Comment (KrzysztofCwalina) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: After all the checks to IsRetained are gone, it seems like the tests are not testing anything.

Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: Are you talking about this particular test or others as well, like MemoryHandleDoubleFree?

Should I remove the Pin test? It is still testing that pinning returns a non-zero pointer address and that Dispose doesn't throw.

**Comment by ahsonkhan** [comment]
> Is anyone even using the IsRetained property anymore? Why do we need it?

The property is used in the OwnedBuffer.Dispose method. We want to throw an exception on Dispose when ref count is greater than 0. How else can we check for double dispose?
